File: src/main/java/org/flossware/civilization/model/MilitaryState.java
package org.flossware.civilization.model;

/**
 * Immutable military state snapshot.
 */
public record MilitaryState(
    long armySize,
    long navySize,
    double techAdvantage,
    double logisticsScore,
    boolean atWar,
    String warOpponent
) {
    public MilitaryState {
        if (armySize < 0) {
            throw new IllegalArgumentException("Army size cannot be negative");
        }
        if (navySize < 0) {
            throw new IllegalArgumentException("Navy size cannot be negative");
        }
        if (logisticsScore < 0 || logisticsScore > 1) {
            throw new IllegalArgumentException("Logistics score must be in [0, 1]");
        }
    }

    public MilitaryState withArmySize(long newSize) {
        return new MilitaryState(newSize, navySize, techAdvantage, logisticsScore, atWar, warOpponent);
    }

    public MilitaryState withNavySize(long newSize) {
        return new MilitaryState(armySize, newSize, techAdvantage, logisticsScore, atWar, warOpponent);
    }

    public MilitaryState withWar(boolean isAtWar, String opponent) {
        return new MilitaryState(armySize, navySize, techAdvantage, logisticsScore, isAtWar, opponent);
    }

    public MilitaryState withTechAdvantage(double advantage) {
        return new MilitaryState(armySize, navySize, advantage, logisticsScore, atWar, warOpponent);
    }
}

File: mobile/shared/src/commonMain/kotlin/org/flossware/civilization/model/MilitaryState.kt
package org.flossware.civilization.model

import kotlinx.serialization.Serializable

@Serializable
data class MilitaryState(
    val armySize: Long,
    val navySize: Long,
    val techAdvantage: Double,
    val logisticsScore: Double,
    val atWar: Boolean,
    val warOpponent: String?
) {
    init {
        require(armySize >= 0) { "Army size cannot be negative" }
        require(navySize >= 0) { "Navy size cannot be negative" }
        require(logisticsScore in 0.0..1.0) { "Logistics score must be in [0, 1]" }
    }

    fun withArmySize(newSize: Long) = copy(armySize = newSize)
    fun withNavySize(newSize: Long) = copy(navySize = newSize)
    fun withWar(isAtWar: Boolean, opponent: String?) = copy(atWar = isAtWar, warOpponent = opponent)
    fun withTechAdvantage(advantage: Double) = copy(techAdvantage = advantage)
}
